sdk/core/Azure.Core.Amqp/src/AmqpAnnotatedMessage.cs

```diff
@@ -110,6 +110,25 @@ public AmqpAnnotatedMessage(AmqpMessageBody body)
         /// Gets or sets the body of the AMQP message.
         /// <seealso href="http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-messaging-v1.0-os.html#type-data"/>
         /// </summary>
-        public AmqpMessageBody Body { get; set; }
+        public AmqpMessageBody Body
+        {
+            get
+            {
+                return _body;
+            }
+            set
+            {
+                // we might be doing pretty advanced trickery to buffer binary body data so we need to make sure to release
+                // the buffers if someone sneaks in new data.
+                if (value != _body && _body.TryGetData(out var bodyData) && bodyData is IDisposable disposable)
```

Comment:
I don't think we should dispose the body we don't own.

```
var message = new AmqpAnnotatedMessage();
message.Body = a;
message.Body = b;
```

Now `a` is disposed and the customer is confused.

Reply:
The fact that anybody can sneak in arbitrary payload even on the receive path is pretty twisted. The whole mutability story here is pretty much hurting any reasonable attempts to do any meaningful internal optimizations...

> I don't think we should dispose the body we don't own.

how is that realistically ever going to happen for a customer scenario? The customer will always have to set an AmqpBody to the body property. The body that is of concern here is publicly exposed as `IEnumerable<ReadOnlyMemory<byte>>`. So in order to be hit by this someone would need to go through the effort of custom wrapping such a custom enumerable and also implement `IDisposable` on that. 

Honestly I see no other way at this stage with the mutable design than doing this unless someone has a better idea. The problem is that on the receive path when the buffer pool is used internally and then someone snoops in new payload over the mutable body property that rented array will never be returned and effectively become a leak. 

Aren't these abstractions anyway just used by EventHub and ServiceBus? If yes why are they even in Core? Another option would be to define a dedicated interface for body releasing inside that Core lib that will be used instead of `IDisposable` because disposable might be too generic, is that the concern?

Reply:
>how is that realistically ever going to happen for a customer scenario? 

What about something like this:

```
var body = message.Body;

message.Body = AmqpMessageBody.FromData(CompressBody(message.Body));
```

Where CompressBody implementation is a lazy IEnumerable.


>Honestly I see no other way

Neither do I currently, but I think that invalidating an object just because it was overwritten in some property of another object after it might've been observed by various parts of the code can lead to some very hard-to-track bugs.

I'll let @JoshLove-msft  comment about the abstraction being in core.

Reply:
I think I have an idea. Let me revert that commit and send in a proposal as a different commit.

Reply:
> ... that will be used instead of IDisposable because disposable might be too generic, is that the concern?

That's where my head was at.  If we can detect out implementation with confidence and avoid accidentally disposing something owned by a caller, I think that mitigates the risk.

Reply:
> I think that mitigates the risk.

I don't think so. The other concern is that someone can hold a reference to the body and that object would be invalidated by the property reassignment.

Reply:
That's fair...   I was focused on the caller-provided scenario and overlooked that.

Reply:
Have a look. I've pushed a different more explicit scope management that early extracts the original body data before it is even exposed to users in the processor event handlers. With that the original body data is released.

Reply:
> Honestly I see no other way at this stage with the mutable design than doing this unless someone has a better idea. The problem is that on the receive path when the buffer pool is used internally and then someone snoops in new payload over the mutable body property that rented array will never be returned and effectively become a leak.

This is a fair point. The body mutability was mostly a holdover from T1, and perhaps we should have not exposed the setter for GA.

Reply:
No not at all - we greatly appreciate all your contributions and it didn't come off as snarky 😄 . In cases where the API leads us to less than ideal perf we definitely want to enhance it where we can.